File: LTHS.py
# License MIT 2024 Ahmad Retha
# Linear Time Hamming Search

import sys
from argparse import ArgumentParser


def report(position, k_mismatches, t, m):
    """ Report (print out) the position, number of mismatches and the pattern found in the text
    position: The start position
    k_mismatches: The number of mismatches
    t: The text
    m: The length of the pattern
    """
    print(f'{position},{k_mismatches}\t{t[position:position+m]}')


def ALPHA(A, c):
    """ Returns the index (position) of letter c in the Alphabet A or -1 if not found
    A: The alphabet
    c: The letter
    """
    return A.find(c)


def initialize_P(A, P, p):
    """ Fills in the Pattern table P in the preprocessing stage
    A: The alphabet
    P: The initialized pattern table
    p: The pattern
    """
    for i, c in enumerate(p):
        j = ALPHA(A, c)
        P[j] = P[j] | (1 << i)


def initialize_T(A, T, t, m):
    """ Fills in the text table T before the search stage proper
    A: The alphabet
    T: The initialized text table
    t: The text
    m: The length of the pattern 
    """
    for i, c in enumerate(t):
        if i >= m-2:
            break
        j = ALPHA(A, c)
        if j == -1:
            continue
        T[j] = T[j] | (1 << i)


def popcount(n):
    """ Returns count of set bits in n
    n: A number
    """
    return n.bit_count()


def search(A, sigma, P, T, t, n, m, k):
    """ Performs the search stage of the LTHS algorithm
    A: The alphabet
    sigma: The size of the alphabet
    P: The pattern table
    T: The text table
    t: The text
    n: The length of the text
    m: The length of the pattern
    k: K-mismatches threshold
    """
    for i in range(m-1, n):
        x = ALPHA(A, t[i])

        if x != -1:
            T[x] = T[x] | (1 << m-1)

        s = 0
        for j in range(sigma):
            diff = (P[j] ^ T[j]) & P[j]
            s = s + popcount(diff)
            T[j] = T[j] >> 1

        if s <= k:
            report(i-m+1, s, t, m)


def LTHS(A, p, k, t):
    """ Performs the Linear Time Hamming Search algorithm
    A: The alphabet, e.g 'ACGT'
    p: The pattern (needle)
    k: K-mismatches threshold (k < |p|)
    t: The text to search (haystack)
    """
    sigma = len(A)
    m = len(p)
    n = len(t)

    P = [0] * sigma
    initialize_P(A, P, p)

    T = [0] * sigma
    initialize_T(A, T, t, m)

    search(A, sigma, P, T, t, n, m, k)


if __name__ == '__main__':
    parser = ArgumentParser(
        prog='LTHS: Linear Time Hamming Search',
        description='Search for a pattern in a text with k-mismatches under the Hamming distance model',
        epilog='License MIT 2024 Ahmad Retha')
    parser.add_argument('-A', '--alphabet', help='Alphabet string, e.g. ACGT', required=True)
    parser.add_argument('-p', '--pattern', help='The pattern (needle)', required=True)
    parser.add_argument('-k', '--k-mismatches', help='K-mismatches threshold', type=int, required=True)
    parser.add_argument('-t', '--text', help='The text to search (haystack)', required=True)
    args = parser.parse_args()

    if len(args.pattern) > len(args.text):
        sys.exit('Pattern is longer than text')

    if args.k_mismatches < 0 or args.k_mismatches >= len(args.pattern):
        sys.exit('Invalid k-mismatches value')

    LTHS(args.alphabet, args.pattern, args.k_mismatches, args.text)

File: lths.c
// License MIT 2024 Ahmad Retha
// LTHS: Linear Time Hamming Search
//
// In this implementation you do not need to provide an alphabet argument to the
// program. It accepts an ASCII pattern and text and uses a lookup array to get
// the index of characters in constant time.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "lths.h"

void printUsage() {
    printf("LTHS: Linear Time Hamming Search\n\n");
    printf("Usage: ./lths <pattern> <k-mismatches> <text>\n");
    printf("Search for a pattern in a text with k-mismatches under the Hamming distance model\n\n");
    printf("arguments:\n");
    printf("\t<PATTERN>\tThe pattern (needle)\n");
    printf("\t<K_MISMATCHES>\tK-mismatches threshold\n");
    printf("\t<TEXT>\t\tThe text to search (haystack)\n\n");
    printf("License MIT 2024 Ahmad Retha\n");
}

void report(char* t, unsigned int i, unsigned int m, unsigned int s) {
    char x[m+1];
    unsigned int pos = i - m + 1;
    strncpy(x, &t[pos], m);
    x[m] = '\0';
    printf("%d,%d\t%s\n", pos, s, x);
}

void search(char* ALPHA, unsigned int sigma, WORD* P, WORD* T, char* t, unsigned int n, unsigned int m, unsigned int k) {
    //do the rest of the positions i=m-1..n-1
    unsigned int c, h, i, j, s;
    for (i = m-1; i < n; i++) {
        c = (unsigned int)t[i];
        j = (unsigned int)ALPHA[c]; //index of letter in T

        if (j != 0) {
            T[j-1] = T[j-1] | (1 << m-1);
        }

        s = 0;
        for (h = 0; h < sigma; h++) {
            WORD diff = (P[h] ^ T[h]) & P[h];
            s = s + PopCount(diff);
            T[h] = T[h] >> 1;
        }

        if (s <= k) {
            report(t, i, m, s);
        }
    }
}

int main(int argc, char** argv) {
    if (argc == 1) {
        printUsage();
        return 0;
    }

    if (argc != 4) {
        printf("Invalid arguments\n\n");
        printUsage();
        return 1;
    }

    unsigned int m = strlen(argv[1]);
    char* p = argv[1];
    unsigned int k = (unsigned int) atoi(argv[2]);
    unsigned int n = strlen(argv[3]);
    char* t = argv[3];

    if (m > n) {
        printf("Error: Pattern is longer than text\n");
        return 1;
    }

    if (m > BITSINWORD) {
        printf("Error: Pattern is too long\n");
        return 1;
    }

    if (k >= m) {
        printf("Error: Invalid k-mismatches value\n");
        return 1;
    }

    //generate ascii alphabet look up table in time O(m+|ascii|)
    //sigma is used for counting the number of unique letters assigned in p
    //ALPHA will store the index of letters not found in p as 0
    char* ALPHA = (char*) calloc(128, sizeof(char));
    unsigned int c, i, j;
    unsigned int sigma = 0;
    for (i = 0; i < m; i++) {
        c = (unsigned int)p[i]; //ascii char number
        if (ALPHA[c] == 0) {
            ALPHA[c] = ++sigma;
        }
    }

    //create pattern table P, requiring space O(|A|) and time O(m+|A|)
    WORD* P = (WORD*) calloc(sigma, WORDSIZE);
    for (i = 0; i < m; i++) {
        c = (unsigned int)p[i]; //ascii char number
        j = (unsigned int)ALPHA[c]; //index of letter in P
        if (j != 0) {
            P[j-1] = P[j-1] | (1 << i);
        }
    }

    //create text table T, requiring space O(|A|) and time O(m+|A|)
    WORD* T = (WORD*) calloc(sigma, WORDSIZE);
    for (i = 0; i < m-1; i++) {
        c = (unsigned int)t[i];
        j = (unsigned int)ALPHA[c]; //index of letter in T
        if (j != 0) {
            T[j-1] = T[j-1] | (1 << i);
        }
    }

    search(ALPHA, sigma, P, T, t, n, m, k);

    //clean up
    free(ALPHA);
    free(P);
    free(T);

    return 0;
}
